package/samplers/carbo/README.md

```diff
@@ -0,0 +1,97 @@
+
+## Installation
+
+```shell
+$ pip install torch --index-url https://download.pytorch.org/whl/cpu
+$ pip install scipy
+```
+
+## Example
+
+```python
+import numpy as np
+import optuna
+import optunahub
+
+
+def objective(trial: optuna.Trial) -> float:
+    x = trial.suggest_float("x", 0.0, 2 * np.pi)
+    y = trial.suggest_float("y", 0.0, 2 * np.pi)
+    c = float(np.sin(x) * np.sin(y) + 0.95)
+    trial.set_user_attr("c", c)
+    return float(np.sin(x) + y)
+
+
+def constraints(trial: optuna.trial.FrozenTrial) -> tuple[float]:
+    c = trial.user_attrs["c"]
+    return (c, )
+
+
+CARBOSampler = optunahub.load_module("samplers/carbo").CARBOSampler
+sampler = CARBOSampler(seed=0, constraints_func=constraints)
+study = optuna.create_study(sampler=sampler)
+study.optimize(objective, n_trials=30)
+
+```
+
+## Others
+
+### Notations
+
+In this section, we use the following notations:
+
+- $x \\in \[0, 1\]^D$, an input vector,
+- $B\_\\epsilon \\coloneqq \[-\\frac{\\epsilon}{2}, \\frac{\\epsilon}{2}\]^D$, an $\\epsilon$-bounding box,
+- $\\xi \\in B\_\\epsilon$, an input noise,
+- $f: \[0, 1\]^D \\rightarrow \\mathbb{R}$, an objective function,
+- $g_c: \[0, 1\]^D \\rightarrow \\mathbb{R}$, the $c$-th constraint,
+- $\\text{LCB}\_{h}: \[0,1\]^D \\rightarrow \\mathbb{R}$, the lower confidence bound of a function $h$,
+- $\\text{UCB}\_{h}: \[0,1\]^D \\rightarrow \\mathbb{R}$, the upper confidence bound of a function $h$.
+
+Suppose we would like to solve the following max-min problem:
+$\\max\_{x \\in \[0,1\]^D} \\min\_{\\xi \\in B\_\\epsilon} f(x + \\xi) \\text{\\ subject \\ to \\ } g_c(x + \\xi) \\geq 0~(\\text{for~}c \\in {1,2,\\dots,C}).$
+where the actual input noise $\\xi$ is assumed to be drawn from $B\_\\epsilon$ uniformly.
+
+### Algorithm Details
+
+1. Train Gaussian process regressors for each function $f, g_1, \\dots, g_C$ using the past observations.
+1. Solve the following max-min problem:
+   $x\_{\\star} \\in \\text{arg}\\max\_{x \\in \[0,1\]^D}\\min\_{\\xi \\in B\_\\epsilon} \\text{UCB}\_{f}(x+\\xi) + \\rho \\sum\_{c=1}^C \[\\text{UCB}\_{g_c}(x+\\xi)\]^{-}$ where $\[a\]^{-} \\coloneqq \\min(0, a)$.
+1. Solve the following minimization problem:
+   $\\xi\_{\\star} \\in \\text{arg}\\min\_{\\xi \\in B\_\\epsilon} \\text{LCB}\_{f}(x\_\\star+\\xi) + \\rho\\sum\_{c=1}^C \[\\text{LCB}\_{g_c}(x\_\\star+\\xi)\]^{-}$
+1. Evaluate each function at $x = x\_{\\star} + \\xi\_{\\star}$.
+1. Go back to 1.
+
+In principle, $\[\\text{UCB}\_{g_c}(x+\\xi)\]^{-}$ and $\[\\text{LCB}\_{g_c}(x+\\xi)\]^{-}$ quantify the upper and lower confidence bounds of the violation amount.
+Please note that Processes 2 and 3 are modified because our setup assumes that the same input noise $\\xi$ is used for each constraint and the objective evaluations.
+Namely, the order of the min or max operation and the summation is flipped in our implementation.
```

Comment:
It is written as "Namely", but since we are talking about a different issue rather than explaining the details of the previous issue, I think "Also" would be better.

```suggestion
Also, the order of the min or max operation and the summation is flipped in our implementation.
```